### PR TITLE
Don't use CriticalSection on MingW without UCRT

### PR DESCRIPTION
**Pull request by khwilliamson**
I ran a test that uses mutexes extensively (and isn't in blead) on MingW, and got lots of failures.  After adding lots of trace statements, and poring over logs, it became clear that the EnterCriticalSection() calls (and kin) were not doing what they are supposed to do, locking out other threads.

This locking mechanism is the default for Windows, but win32thread.h allows for the use of a heavier-weight locking mechanism, selected by a Configuration option, or by adding to it the single line

  #define DONT_USE_CRITICAL_SECTION

Doing so "magically" caused those errors to go away.

This tells me that the error is in the CriticalSection libc calls.

I then ran the same test on MSVC, and the errors didn't happen.  It is using a later libc that the MingW I tested on is.  From this, I infer that this bug has been fixed in later libc versions.  (This is true of some other bugs we already have workarounds for.)

This commit causes the CriticalSection calls to not be used on MingW when it isn't compiled to use the later libc version UCRT.  I don't know when libc was fixed, but until evidence to the contrary surfaces, I took the easy way out, and assume any version is ok that is recent enough to have _MSVC_VER defined .

### CODE REVIEW

**Comment by jkeenan** [comment]
Is this a blocker for the release of 5.40?

Can we get a unit test to exercise this code?

**Comment by tonycoz** [comment]
Win32 critical sections are part of the Win32 API and have nothing to do with which libc is in use.

They do have some limitations compared to mutexes in that they can't be safely copied or moved, but I don't see any evidence from a quick grep that we do that.

**Comment by xenu** [comment]
Yeah, I don't see a reason why the behaviour would differ between the compilers.

Do you have a reproducer?

**Comment by khwilliamson** [comment]
`git checkout smoke-me/khw-env`

If you compile on MingW without UCRT and run `lib/locale_threads.t`, and capture the output you will get some errors.  If you recompile after uncommenting this line

`//#define DONT_USE_CRITICAL_SECTION`

in `win32/win32threads.h`, and re-run that test file, you will typically get just one error.  I'm unsure if that error is mine or the system's, though I'm leaning towards the latter.  As I said in the issue description, by turning on the tracing statements I added and analysing the output, I determined that the `EnterCriticalSection` calls weren't locking out other threads.

Note that the test file is modified in this branch to just have 3 threads running

**Comment by tonycoz** [comment]
I've reproduced this and spent quite a lot of time trying to work it out.

There's pretty much no chance there's a bug in critical sections, they're been part of the Win32 API since the beginning, if they were broken they would have been fixed long ago.

From looking at the test failures I saw, it looked like there was a mismatch in the way characters were being encoded, the expected value would be full of extended characters, which were replaced in the failed result with `"?"`.

After spending way too much time attack this from the synchronization side, I tried looking at the character encoding.

 tried a very hacky change in S_strftime_tm:
```
#ifdef __MINGW32__
        /* hacky */
        wchar_t fmttmp[200];
        wchar_t tmp[200];
        _snwprintf(fmttmp, 200, L"%S", fmt);
        Size_t wlen = wcsftime(tmp, 200, fmttmp, mytm);
        Size_t len = snprintf(buf, bufsize, "%S", tmp);
#else
        Size_t len = strftime(buf, bufsize, fmt, mytm);
#endif
```
and `locale_threads.t` passes.

**Comment by khwilliamson** [comment]
tl;dr  we don't fully understand this issue, but we are considering it moot because it doesn't happen when using UCRT, so I'm closing this PR

I chatted with @tonycoz, and we came to the following conclusions (as I see it)

There is something weird going on here.  He says that bugs in the mutex handling would create havoc with all sorts of applications and would have been fixed long ago.  But I examining the traces I made saw that a thread started executing when another thread had a lock on the CPU.

This comes from code that is not in blead.  The question marks are to be expected because what's happening is that the global locale is getting reset to something the strftime isn't expecting, and so substitutes the question mark instead of trying to return an illegal sequence.  The locking is supposed to make sure strftime only runs when the locale is consistent with what it is expecting.  It's unclear why wcsftime works.  Perhaps it has a different timing; perhaps strftime is buggy.

But it all works when using thread-safe locales and not the global one.  That indicates the underlying problem is not strftime vs wcsftime.

So the issue goes away when compiled with UCRT, as that automatically selects thread-safe locales.  Someone could still supply a flag to Configure to defeat this, but that is unlikely, and the solution would be to just not use that flag.

Both Tony and I have spent way too much time on this, so don't see a need to pursue it further.  Neither of us can find a bug related to this in the perl implementation.  UCRT should be used because there are a bunch of known bugs that it fixes.  Strawberry looks to be furnishing a UCRT MingW, so the solution is to upgrade to a known better libc.  And this problem goes away, along with others.
